devtools/gradle/src/main/java/io/quarkus/gradle/builder/QuarkusModelBuilder.java

```diff
@@ -0,0 +1,285 @@
+        }
+
+        final Set<Dependency> extensionDependencies = collectExtensionDependencies(project, directExtensionDependencies);
+
+        ArtifactCoords appArtifactCoords = new ArtifactCoordsImpl(project.getGroup().toString(), project.getName(),
+                project.getVersion().toString());
+
+        return new QuarkusModelImpl(getWorkspace(project),
+                getAdditionalWorkspaces(project.getRootProject(), project.getName()),
+                appDependencies,
+                extensionDependencies);
+    }
+
+    private Workspace getWorkspace(Project project) {
+        ArtifactCoords appArtifactCoords = new ArtifactCoordsImpl(project.getGroup().toString(), project.getName(),
+                project.getVersion().toString());
+        final SourceSet mainSourceSet = QuarkusGradleUtils.getSourceSet(project, SourceSet.MAIN_SOURCE_SET_NAME);
+
+        return new WorkspaceImpl(appArtifactCoords, project.getProjectDir().getAbsoluteFile(),
+                project.getBuildDir().getAbsoluteFile(), getSourceSourceSet(mainSourceSet), convert(mainSourceSet));
+    }
+
+    private Set<Workspace> getAdditionalWorkspaces(Project project, String mainProjectName) {
+        Set<Workspace> workspaces = new HashSet<>();
+        for (Project subproject : project.getSubprojects()) {
+
+            if (subproject.getName().equals(mainProjectName)) {
+                continue;
+            }
+            final Convention convention = subproject.getConvention();
+            JavaPluginConvention javaConvention = convention.findPlugin(JavaPluginConvention.class);
+            if (javaConvention == null) {
+                continue;
+            }
+
+            workspaces.add(getWorkspace(subproject));
+        }
+        return workspaces;
+    }
+
+    private Set<org.gradle.api.artifacts.Dependency> getDirectExtension(Project project) {
+        final Set<org.gradle.api.artifacts.Dependency> directExtension = new HashSet<>();
+        // collect enforced platforms
+        final Configuration impl = project.getConfigurations()
+                .getByName(JavaPlugin.IMPLEMENTATION_CONFIGURATION_NAME);
+        for (org.gradle.api.artifacts.Dependency d : impl.getAllDependencies()) {
+            if (!(d instanceof ModuleDependency)) {
+                continue;
+            }
+            final ModuleDependency module = (ModuleDependency) d;
+            final Category category = module.getAttributes().getAttribute(Category.CATEGORY_ATTRIBUTE);
+            if (category != null && Category.ENFORCED_PLATFORM.equals(category.getName())) {
+                directExtension.add(d);
+            }
+        }
+        return directExtension;
+    }
+
+    private Set<org.gradle.api.artifacts.Dependency> getDirectExtensionDependencies(Set<ResolvedDependency> dependencies,
+            Collection<Dependency> appDependencies) {
```

Comment:
1) this method is supposed to collect only the first met extension artifacts (could be direct or transitive but the first met walking the dependency tree;
2) in the original code I was also checking whether I have already visited some deps (that could happen);
3) just fyi, we try to stay away from lambda expressions for now for performance reasons.